src/software/backend/simulator_backend.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <mutex>
+
+#include "software/ai/world/world.h"
+#include "software/backend/backend.h"
+
+class SimulatorBackend : public Backend
+{
+   public:
+    static const std::string name;
+
+    /**
+     * Creates a new SimulatorBackend
+     *
+     * The SimulatorBackend will run a physics simulation to simulate and publish the
+     * World, and simulate the primitives given to it. The SimulatorBackend will only
+     * advance the simulation by a fixed amount each time primitives are received, rather
+     * than running at full-speed or in real-time. This is because we want the simulator
```

Comment:
Is the SimulatorBackend's "world" different from the physics engine's world? When I say the SimulatorBackend's "world" I mean the thing that is updated in "updateSimulation".

If this isn't the case, why wouldn't we want to time step the physics engine in real time. The physics engine only cares about the physics of the world at its current state (ie doesn't care about any primitives or ai), it shouldn't care about previous or future states.

Reply:
Right now the current SimulatorBackend's `world` is the AI representation of the world using our `World` class. There will be a separate datastructure maintained by the physics engine that we will need to translate into our `World` datatype when we want to expose the current simulation state.

The reason we don't want to run in real-time is because the simulation can likely run much faster than real life. This is a problem because if the simulation can run at 300Hz (for example), but the AI can only run at 30Hz, the buffers between the Observers will overflow and data will be lost. This is not realistic because generally the input runs at 60Hz (roughly the same rate as the camera fps) and the AI runs at about 30Hz.

This means that IRL each time the AI is ready for more data, about 2 * 1/60 seconds have passed in real-time (and therefore `World` time). But if the simulation is running way faster then 10 * 1/60 seconds will have passed each time the AI is ready for more data.

And generally it's not recommended to update the physics engine with different time steps each time, so using a "dynamic" step in the simulation while trying to run as fast as possible isn't an ideal solution.

Reply:
I've update the code comment to be more explicit about this.

Reply:
That makes sense

What I'm worried about is if we choose some arbitrary constant number of steps per primitive update it will be out of sync at certain time intervals. Like, it won't account for the speed up or slow down of the AI.

For example, if we choose some constant K as the num_steps_per_primitive_update, and we chose it in the ideal case of the AI running at 30Hz. If the AI starts running at 20Hz I think the simulator world will be out of sync with the AI world. Since we chose K to be ideal for 30Hz so it will perform K steps expecting an update every "30Hz in seconds".

Reply:
That is a good point. If the relative speeds change significantly that "ratio" would have to be changed. I'm open to ideas but our main concern right now is determinism. We may need to change things / add ways it can run if we want to support other features (like running and showing tests within the Visualizer).

If we can't think of anything for now i'd like to move forwards with this just to get an MVP done so we can write Play tests, but it's something we should be aware of for sure.

Reply:
Sounds good